File: 343. 整数拆分/main.java
class Solution {
    public int integerBreak(int n) {
        if (n == 2) {
            return 1;
        } else if (n == 3) {
            return 2;
        } else if (n % 3 == 0) {
            return f_pow(3, n / 3);
        } else if (n % 3 == 1) {
            return 2 * 2 * f_pow(3, (n-4) / 3);
        } else {
            return 2 * f_pow(3, (n-2) / 3);
        }
    }
    
    private int f_pow(int a, int b) {
        int ans = 1;
        while (b > 0) {
            if (b % 2 == 1) ans *= a;
            a *= a;
            b >>= 1;
        }
        return ans;
    }
}

File: 343. 整数拆分/main.py
# 用数学归纳法证明对任意一个整数最大的拆分形式为 n = 2^p * 3^q (n >= 3, p, q为自然数)，
# 并且q应该尽可能大（理由：如果q减少2，则p要增加3，而2^3 < 3^2，因此应该尽可能增加q）。

class Solution:
    def integerBreak(self, n: int) -> int:
        if n <= 3: return n - 1
        a, b = n // 3, n % 3
        if b == 0:
            return int(math.pow(3, a))
        if b == 1:
            return int(math.pow(3, a - 1) * 4)
        return int(math.pow(3, a) * 2)
